lib/kino/js/live.ex

```diff
@@ -0,0 +1,369 @@
+
+  Whenever the widget is rendered on the client, the `c:handle_connect/1`
+  callback is called and it builds the initial data for the
+  client. In this case, we always return the stored `html`.
+  This initial data is then passed to the JavaScript `init`
+  function. Keep in mind that while the server is initialized
+  once, connect may happen at any point, as the users join/refresh
+  the page.
+
+  Finally, the whole point of our example is the ability to
+  replace the HTML content directly from the Elixir side and
+  for this purpose we added the public `replace(widget, html)`
+  function. Underneath the function uses `cast/2` to message
+  our server and the message is handled with `c:handle_cast/2`.
+  In this case we store the new `html` in the server state and
+  broadcast an event with the new value. On the client side,
+  we subscribe to those events with `ctx.handleEvent(event, callback)`
+  to update the page accordingly.
+
+  ## Event handlers
+
+  You must eventually register JavaScript handlers for all events
+  that the client may receive. However, the registration can be
+  deferred, if the initialization is asynchronous. For example,
+  the following is perfectly fine:
+
+  ```js
+  export function init(ctx, data) {
+    fetch(data.someUrl).then((resp) => {
+      ctx.handleEvent("update", (payload) => {
+        // ...
+      });
+    });
+  }
+  ```
+
+  Or alternatively:
+
+  ```js
+  export async function init(ctx, data) {
+    const response = await fetch(data.someUrl);
+
+    ctx.handleEvent("update", (payload) => {
+      // ...
+    });
+  }
+  ```
+
+  In such case all incoming events are buffered and dispatched once
+  the handler is registered.
+  '''
+
+  defstruct [:module, :pid]
+
+  @doc false
+  use GenServer
+
+  require Logger
+
+  alias Kino.JS.Live.Context
+
+  @type t :: %__MODULE__{module: module(), pid: pid()}
+
+  @doc """
+  Invoked when the widget server started.
+
+  See `c:GenServer.init/1` for more details.
+  """
+  @callback init(arg :: term(), ctx :: Context.t()) :: {:ok, ctx :: Context.t()}
+
+  @doc """
+  Invoked whenever a new client connects to the server.
+
+  The returned data is passed to the JavaScript `init` function
+  of the connecting client.
+  """
+  @callback handle_connect(ctx :: Context.t()) :: {:ok, data :: term(), ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle client events.
+  """
+  @callback handle_event(event :: String.t(), payload :: term(), ctx :: Context.t()) ::
+              {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle asynchronous `cast/2` messages.
+
+  See `c:GenServer.handle_cast/2` for more details.
+  """
+  @callback handle_cast(msg :: term(), ctx :: Context.t()) :: {:noreply, ctx :: Context.t()}
+
+  @doc """
+  Invoked to handle synchronous `call/3` messages.
+
+  See `c:GenServer.handle_call/3` for more details.
+  """
+  @callback handle_call(msg :: term(), {pid(), reference()}, ctx :: Context.t()) ::
```

Comment:
It is actually not guaranteed to be `{pid(), reference()}`. For example, in some cases it is `[alias | reference()]`. This should be an opaque term:

```suggestion
  @callback handle_call(msg :: term(), term(), ctx :: Context.t()) ::
```